### PR TITLE
Create package coq-runtime-types

### LINKED ISSUE #15936: Package uint63.ml and friends separately
From the manual (emphasis mine):
> The extraction of these primitives can be customized similarly to the extraction of regular axioms (see Program extraction). Nonetheless, the ExtrOCamlInt63 module can be used when extracting to OCaml: it maps the Coq primitives to types and functions of a Uint63 module (including signed functions for Sint63 despite the name). That OCaml module is not produced by extraction. Instead, it has to be provided by the user (if they want to compile or execute the extracted code). For instance, **an implementation of this module can be taken from the kernel of Coq.**

In the case of primitive integers, the relevant files are `uint63.mli` and the two implementations `uint63_32.ml` and `uint63_63.ml`. There are also relevant files for primitive floats and arrays.

Anyone who extracts primitive objects from Coq and wishes to compile them would benefit from using these files from the Coq kernel. It would be even better if these files lived in their own opam package that could be installed without installing Coq.

There has already been some discussion [on Zulip](https://coq.zulipchat.com/#narrow/stream/237663-coq-community-devs-.26-users/topic/Advising.20against.20licenses.3F/near/279542086) about this. Some problems that were mentioned with the current state of things:
- `uint63.ml` is compiled with `-rectypes`, which is a problem for coq-simple-io (https://github.com/Lysxia/coq-simple-io/pull/50)
- The licensing situation is unclear, which was the [initial](https://coq.zulipchat.com/#narrow/stream/237663-coq-community-devs-.26-users/topic/Advising.20against.20licenses.3F/near/279029933) [catalyst](https://coq.zulipchat.com/#narrow/stream/237663-coq-community-devs-.26-users/topic/Advising.20against.20licenses.3F/near/279540148) for the discussion, even though the consensus seems to be that [there is no actual problem](https://coq.zulipchat.com/#narrow/stream/237663-coq-community-devs-.26-users/topic/Advising.20against.20licenses.3F/near/279541937)

### PR DESCRIPTION
**Pull request by Lysxia**
Closes #15936 

- [x] Added **changelog**.
- [x] Added / updated **documentation**.
- [ ] Opened **overlay** pull requests (hopefully none necessary).

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

File: primitive/dune [hunk 1]
```diff
@@ -0,0 +1,25 @@
+(library
+ (name primitive)
+ (synopsis "Coq Primitive Objects")
+ (public_name coq-primitive)
+ (wrapped false)
+ (flags (:standard -no-rectypes))
```
Comment (Lysxia) [inline primitive/dune:6]: Is there a better way to disable `rectypes`? My understanding is that `:standard` inherits from the `(env)` in the `dune` file at the root.

Comment (Alizter) [reply primitive/dune:6]: You can do `:standard \ -rectypes` perhaps?

File: doc/sphinx/language/core/primitive.rst [hunk 2]
```diff
@@ -53,11 +53,8 @@ applications of these primitive operations.
 The extraction of these primitives can be customized similarly to the extraction
 of regular axioms (see :ref:`extraction`). Nonetheless, the :g:`ExtrOCamlInt63`
 module can be used when extracting to OCaml: it maps the Coq primitives to types
-and functions of a :g:`Uint63` module (including signed functions for
-:g:`Sint63` despite the name). That OCaml module is not produced by extraction.
-Instead, it has to be provided by the user (if they want to compile or execute
-the extracted code). For instance, an implementation of this module can be taken
-from the kernel of Coq.
+and functions of the :g:`Uint63` module from the `coq-primitive` library
+(including signed functions for :g:`Sint63` despite the name).
```
Comment (jfehrle) [inline doc/sphinx/language/core/primitive.rst:57]: I don't quite see why you say "despite the name".  What's the apparent inconsistency?

File: doc/changelog/11-infrastructure-and-dependencies/16094-master.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+- **Added:**
+  New `coq-primitive` library, packaging the implementation of Coq primitive objects
+  (integers, floats, arrays) for use by extracted code.
```
Comment (jfehrle) [inline doc/changelog/11-infrastructure-and-dependencies/16094-master.rst:3]:
```suggestion
  A `coq-primitive` library, packaging the implementation of Coq primitive objects
  (integers, floats, arrays) for use by extracted code
```

"New" seems redundant with "Added".
No period on the end of the line

**Review by jfehrle — APPROVED** [review]
Very minor comments.

**Comment by proux01** [comment]
There seems to be a missing `coq-core` -> `coq-primitive` dependency.

File: doc/sphinx/language/core/primitive.rst (hunk 2 above)
Comment (ana-borges) [reply doc/sphinx/language/core/primitive.rst:57]: The OCaml module for primitive ints is called `Uint63`, so it looks like it should contain only unsigned things.

**Comment by ana-borges** [comment]
Awesome, thanks for doing this!

Should there be a part of the refman describing the new package? Or is there somewhere else where that kind of documentation should live?

**Comment by Lysxia** [comment]
There are also some C files to be included for the `external` definitions. BTW Why are there both `float64_31` and `float64_63`? Aren't OCaml `float` guaranteed to be 64 bit?

> Should there be a part of the refman describing the new package? Or is there somewhere else where that kind of documentation should live?

Good question. I'm not sure of the answer.

File: kernel/dune [hunk 4]
```diff
@@ -3,8 +3,8 @@
  (synopsis "The Coq Kernel")
  (public_name coq-core.kernel)
  (wrapped false)
- (modules (:standard \ genOpcodeFiles uint63_31 uint63_63 float64_31 float64_63))
- (libraries boot lib coqrun dynlink))
+ (modules (:standard \ genOpcodeFiles))
+ (libraries boot lib coqrun dynlink primitive))
```
Comment (Alizter) [inline kernel/dune:7]: @proux01 This is where "coq-core" depends on "coq-primitive".

**Comment by proux01** [comment]
> Why are there both `float64_31` and `float64_63`? Aren't OCaml `float` guaranteed to be 64 bit?

Sure, floats are always 64 bits, but on old 32 bits systems, there is a risk of double rounding which prevents directly using OCaml operators and requires the use of a C implementation.

File: kernel/dune (hunk 4 above)
Comment (proux01) [reply kernel/dune:7]: That doesn't seem enough according to the CI.

Comment (Alizter) [reply kernel/dune:7]: Right, in the dune-project file the package for coq-core also needs to depend on coq-primitive. Look at how coq-stdlib relies on coq-core.

**Comment by Lysxia** [comment]
I'm not sure I packaged the C code properly yet. How do I solve ```undefined reference to `coq_uint63_to_float'``` when building an executable using coq-primitive?

**Comment by proux01** [comment]
The fact that some C files are duplicated (but not all) indeed looks really bad. You should probably not copy anything and tell dune to reuse the same C files in primitive and the VM. You can maybe keep everything in `kernel` rather than creating a new `primitive` directory, that might be easier (or not, my dune skills are almost inexistant).

**Comment by Lysxia** [comment]
There isn't much duplication. It's mainly moving `coq_float64.c` and two C functions to implement `Uint63.to_float` that are not used anywhere else. There are just two macros being copied, which are really general OCaml FFI idioms rather than things specific to Coq kernel.

Moving things in a separate directory works better dune's default behavior of including everything in a directory, otherwise modules have to explicitly be excluded from one of the libraries sharing the same directory. Furthermore, the main intended consumers after the kernel are extraction users, so this package is intentionally kernel-agnostic.

**Comment by proux01** [comment]
Indeed, my bad.
You should probably take the opportunity to move the float <-> int63 conversion functions from uint63_{31,63}.ml to float64_{31,63}.ml, they were put there for historical reasons that no longer make sense today. This would be more in line with the C file.

Also note that https://github.com/coq/coq/pull/15560 will also modify build system which may conflict a bit with the current PR (hopefully not much as it mostly touches coq-stdlib).

More importantly, note that the `coq_float64.c` file should only be compiled after running the `compute_cflags` and `check_fmath` functions of `configure.ml`.

File: coq-primitive.opam [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# This file is generated by dune, edit dune-project instead
+opam-version: "2.0"
+version: "dev"
+synopsis: "The Coq Proof Assistant -- Primitive Objects"
+description: """
+This package provides OCaml implementations of Coq's primitive
+objects, used by coq-core and also available to link with extracted code."""
+maintainer: ["The Coq development team <coqdev@inria.fr>"]
+authors: ["The Coq development team, INRIA, CNRS, and contributors"]
+license: "LGPL-2.1-only"
```
Comment (Zimmi48) [inline coq-primitive.opam:10]: I thought that we had discussed making this package MIT licensed (and that you had collected all the necessary authorizations from the relevant contributors). If that's the case, then the headers of the files contained in the packages should be adapted as well.

File: primitive/coq_float64.c [hunk 6]
```diff
@@ -72,3 +74,13 @@ DECLARE_FUNOP(next_down, next_down(x))
...
+
+#define uint63_of_value(val) ((uint64_t)(val) >> 1)
+
+double coq_uint63_to_float(value x) {
+  return (double) uint63_of_value(x);
+}
+
+value coq_uint63_to_float_byte(value x) {
+  return caml_copy_double(coq_uint63_to_float(x));
+}
```
Comment (proux01) [inline primitive/coq_float64.c:86]: I really don't know why I wrote this like that but the easiest solution is to remove that C code and use the OCaml primitive `float_of_int` in `uint63_63.ml`, that is replace there the following lines
```OCaml
external to_float : int -> (float [@unboxed])
  = "coq_uint63_to_float_byte" "coq_uint63_to_float"
[@@noalloc]
```
by
```OCaml
let to_float = float_of_int
```
Sorry about that

File: primitive/coq_float64.c [hunk 7]
```diff
@@ -4,7 +4,7 @@
 /* <O___,, * (see version control and CREDITS file for authors & dates) */
 /*   \VV/  **************************************************************/
 /*    //   *    This file is distributed under the terms of the         */
-/*         *     GNU Lesser General Public License Version 2.1          */
+/*         *                     MIT License                            */
```
Comment (proux01) [inline primitive/coq_float64.c:7]: Shouldn't the file be dual licensed?

File: kernel/byterun/dune [hunk 8]
```diff
@@ -4,7 +4,7 @@
  (public_name coq-core.vm)
  (foreign_stubs
   (language c)
-  (names coq_fix_code coq_float64 coq_memory coq_values coq_interp)
+  (names coq_fix_code coq_memory coq_values coq_interp)
```
Comment (proux01) [inline kernel/byterun/dune:7]: `coq_float64` is used in the VM:
```
Error: Error on dynamically loaded library: ./kernel/byterun/dllcoqrun_stubs.so: ./kernel/byterun/dllcoqrun_stubs.so: undefined symbol: coq_next_up
```

File: primitive/float64_31.ml [hunk 9]
```diff
@@ -4,7 +4,7 @@
 (* <O___,, * (see version control and CREDITS file for authors & dates) *)
 (*   \VV/  **************************************************************)
 (*    //   *    This file is distributed under the terms of the         *)
-(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *                     MIT License                            *)
```
Comment (proux01) [inline primitive/float64_31.ml:7]: Are you sure you got agreements of all contributors to relicense that file? (personally, I only agreed about uint63{,_31,_63}.ml{,i}, not that I would disagree though). Same for all float and array files.

Comment (Lysxia) [reply primitive/float64_31.ml:7]: Ah sorry you're right, I only asked about uint63... Should I do another round of emails, or just ping everyone here?

File: kernel/byterun/dune (hunk 8 above)
Comment (proux01) [reply kernel/byterun/dune:7]: Sure, I just mean that the VM (kernel/byterun/coq_interp.c) needs to somehow have access to the content of the C file. I'm unfortunately not fluent enough in dune to tell you the best way to achieve this.

Comment (ejgallego) [reply kernel/byterun/dune:7]: That's a linking error, not a compile-time one.

**Comment by ejgallego** [comment]
Yeah I find just primitive a bit bare, but as always with naming, it is not easy. Feel free folks to choose the name you want, I'm only suggesting ideas. The opam package could maybe be `coq-runtime-number-types` or something like that?

**Comment by ejgallego** [comment]
IMVHO `runtime` for the opam package makes sense as we expect packages to install only that one, not the full Coq.

**Comment by proux01** [comment]
Maybe split into `coq-runtime-number-types` and `coq-runtime-array-type` to avoid persistent arrays to be in something named "number"?

**Comment by Lysxia** [comment]
@Sakarah @maximedenes  @silene  @proux01 @ppedrot  @erikmd This PR relicenses the implementation of primitive ints, floats, and arrays, from LGPL-2.1-only to MIT. Those files are recent enough that their contributors can be enumerated easily. (I previously asked some of you about uint63 only.) May I have your permission for this relicensing?

The goal for this repackaging is to make it easier to build extracted code, notably by not depending on the whole `kernel` library. The LGPL has some subtleties about redistribution (mainly, if you bundle the files somewhere, they must remain under LGPL, but it doesn't infect the rest, [if you trust our local legal non-experts](https://coq.zulipchat.com/#narrow/stream/237663-coq-community-devs-.26-users/topic/Advising.20against.20licenses.3F)). MIT would make the situation a no-brainer.

File: primitive/coq_float64.c (hunk 7 above)
Comment (Lysxia) [reply primitive/coq_float64.c:7]: Why dual license it?

File: runtime/LICENSE [hunk 10]
```diff
@@ -0,0 +1,19 @@
+Copyright © 2022 INRIA, CNRS and contributors
```
Comment (Zimmi48) [inline runtime/LICENSE:1]: Just do not put any year. I know it's not conventional, but there is no legal obligation to follow the convention here. No year means no update ever necessary.

**Comment by ppedrot** [comment]
@Lysxia not sure I ever wrote anything meaningful in those files, but you have my approval for MIT relicensing.

**Comment by Zimmi48** [comment]
> Under US law, copyrights can't last forever, for example copyrighted books eventually lose protection and enter the public domain.

@jfehrle That's true outside US too. Copyright laws are quite normalized nowadays, even though some subtle differences remain.

>Interesting not to have a year on a copyright notice.

Copyright notices are not a legal requirement anymore. They used to be important to get protection in the US before the US entered the [Bernes convention](https://en.wikipedia.org/wiki/Berne_Convention) in 1989. Since they are now purely informative, it's fine to deviate from the conventional presentation. That being said, I'm fine with including a date, as long as we agree that we won't update it.

**Comment by Zimmi48** [comment]
>I do not agree to the relicensing of my code to a MIT license.

That settles things then :disappointed: I'm guessing you already have the various consequences well in mind and therefore that there is no point in arguing?

**Comment by silene** [comment]
Yes, the whole point of creating this package is so that no relicensing of Coq's code is needed in the first place anyway.

**Comment by ppedrot** [comment]
I agree that at the very least we should adopt the OCaml linking exception for the license of this package.

**Comment by Zimmi48** [comment]
My reason for preferring the MIT license is just that it is much simpler (to read, to understand, to abide by), but the OCaml-LGPL-linking-exception option is fine as well.

**Comment by erikmd** [comment]
> This PR relicenses the implementation of primitive ints, floats, and arrays, from LGPL-2.1-only to MIT.

I agree to the relicensing to the MIT license. Otherwise, I'm also fine with the LGPL + OCaml linking exception.

**Comment by proux01** [comment]
@Lysxia I made the CI a lot "greener" but you still have a few jobs to fix (plugins and test-suite)
Note that this definitely conflicts with #15560 so it's probably better to have #15560 merged first and then rebase this on top of it.

**Comment by ejgallego** [comment]
> The dune build seems to disable native compilation, for a good reason I assume?

Just to make the default build faster.

I see the problem, shouldn't be hard to fix.

**Comment by Alizter** [comment]
@Lysxia Can you rebase this?

**Comment by proux01** [comment]
According to the CI, compilation seems broken?

**Comment by Lysxia** [comment]
Oops, sorry about that. There will still be failures later in the pipeline due to native compilation not yet knowing where to look (see details above).

**Comment by Lysxia** [comment]
Indeed `-rectypes` takes away the main motivation, so I will close this.

I would still like to have a dedicated package for these files so that extracted code using primitive types can be built without all of `coq`.  A simpler solution for that seems to be to copy the relevant files to a different repository. This means some code duplication, but I don't imagine there will be much change to these files, and it seems a lesser evil compared to complexifying the setup of coq. That should be legal as long as the LGPL license is preserved, right?

**Comment by Zimmi48** [comment]
> That should be legal as long as the LGPL license is preserved, right?

Yes.

>it seems a lesser evil compared to complexifying the setup of coq

Since Coq is already split into multiple packages nowadays, I am not sure anyone would view this as complexifying the setup.

**Pull request closed without merging**